ckan/config/middleware/flask_app.py

```diff
@@ -251,55 +288,78 @@ def make_flask_stack(conf: Union[Config, CKANConfig]) -> CKANApp:
-
-        return user
+    login_manager.user_loader(load_user)
+    login_manager.request_loader(load_user_from_request)
 
     # Update the main CKAN config object with the Flask specific keys
     # that were set here or autogenerated
     flask_config_keys = set(flask_app.config.keys()) - set(config.keys())
     for key in flask_config_keys:
         config[key] = flask_app.config[key]
 
-    # Prevent the host from request to be added to the new header location.
-    app = HostHeaderMiddleware(app)
-
-    app = I18nMiddleware(app)
+    # Prevent the host from request to be added to the new header
+    # location. Here `wsgi_app` can be missing if IMiddleware plugin replaces
+    # app instead of adding before/after request callbacks.
+    app.wsgi_app = HostHeaderMiddleware(getattr(app, "wsgi_app", app))
+    app.wsgi_app = I18nMiddleware(app.wsgi_app)
 
-    # Add a reference to the actual Flask app so it's easier to access
-    # type_ignore_reason: custom attribute
-    app._wsgi_app = flask_app  # type: ignore
+    # Add a reference to the actual Flask app so it's easier to access. It can
+    # be not required as CKAN does not replaces Flask app starting from
+    # v2.12. But some extensions still can do it due to incorrect
+    # implementation of IMiddleware.
+    setattr(app, "_wsgi_app", flask_app)
 
     return app
 
 
+def load_user(user_id: str) -> model.User | model.AnonymousUser | None:
+    """
+    This callback function is called whenever we need to reload from
+    the database the logged in user in the session (ie the cookie).
+
+    Site maintainers can choose to completely ignore cookie based
+    authentication for API calls, but that will break existing JS widgets
+    that rely on API calls so it should be used with caution.
+    """
+    for p in PluginImplementations(IAuthenticator):
+        if user := p.identify_session_user(user_id):
+            return user
+
+    endpoint = request.endpoint or ""
+    is_api = endpoint.split(".")[0] == "api"
+    if is_api and not config.get("ckan.auth.enable_cookie_auth_in_api"):
+        return
+    return model.User.get(user_id)
+
+
+def load_user_from_request(
+        request: Request
+) -> model.User | model.AnonymousUser | None:
+    """
+    This callback function is called whenever a user could not be
+    authenticated via the session cookie, so we fall back to the API token.
+    """
+    for p in PluginImplementations(IAuthenticator):
+        if user := p.identify_request_user(request):
+            return user
+
+    g.login_via_auth_header = True
+
+    return _get_user_for_apitoken()
+
+
+def _get_user_for_apitoken() -> model.User | None:
+    apitoken_header_name = config["apitoken_header_name"]
+    apitoken: str = request.headers.get(apitoken_header_name, "")
+
+    if not apitoken:
+        return None
+    apitoken = str(apitoken)
+    log.debug('Received API Token: %s[...]', apitoken[:10])
+
+    return api_token.get_user_from_token(apitoken)
```

Comment:
it'd be useful to log the user name (if one is found) at the same time as the token